# Toronto Neighborhoods

### First Section!

We start by importing the packages

In [1]:
import pandas as pd

Then we import the table from wikipedia

In [2]:
link="https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
tables=pd.read_html(link)
db=tables[0]
db

Postal code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                          Neighborhood  
0                                                  NaN  
1                                                  NaN  
2                                            Parkwoods  
3                                     Victoria Village  
4                           Regent Park / Harbourfront  
..                                                 ...  
175                                                NaN  
176                                                NaN  
177                                                NaN  
178  Mimico NW / The Queensway West / South of Bloo...  
179                                                NaN  

[180 rows x 3 columns]

Then, we eliminate the rows that dont have an assigned Borough

In [3]:
indexNames = db[db['Borough'] == "Not assigned"].index
indexNames
db=db.drop(indexNames)
db=db.reset_index(drop=True)
db

Postal code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                           Regent Park / Harbourfront  
3                    Lawrence Manor / Lawrence Heights  
4         Queen's Park / Ontario Provincial Government  
..                                                 ...  
98     The Kingsway / Montgomery Road / Old Mill North  
99                                Church and Wellesley  
100              Business reply mail Processing CentrE  
101  Old Mill South / King's Mill Park / Sunnylea /...  
102  Mimico NW / The Queensway West / South of Bloo...  

[103 rows x 3 columns]

To check if there are Neighborhoods that are not assigned

In [4]:
indexNames = db[ db['Neighborhood'] == "Not assigned" ].index
indexNames.shape

(0,)

Since there are no Neighborhoods not assigned, we will continue to concatenate the data for the cases where a postal code includes several neighborhoods.

In [5]:
dupPC =db[db["Postal code"].duplicated()]
dupPC.shape

(0, 3)

We found 0 duplicated postal codes. It seems the original table from wikipedia grouped the different neighborhoods from a single postal code. Therefore, the data frame needs no further changes.

Finally, lets find the size of the dataframe

In [6]:
db.shape

(103, 3)

### Second Section!

To search for the location of each neighborhood we need to import the package

In [7]:
import geocoder 

Then, we search for the location of each neighborhood

In [8]:
d=[]
nonfound=0
n=1 #Number of iterations, change if API key is found
for i in range(len(db)) :
    cont=0
    lat_lng_coords = None
    postal_code= db.at[i,'Postal code']
    while(lat_lng_coords is None and cont<n): 
        a='{}, Toronto, Ontario'.format(postal_code)
        g = geocoder.google(a)
        print(g)
        lat_lng_coords = g.latlng
        cont=cont+1
    if cont==n:
        nonfound=nonfound+1
    else:
        latitude = lat_lng_coords[0]
        longitude = lat_lng_coords[1]
        d.append(
            {
                'lat': latitud,
                'long': longitude
            }
        )
print(nonfound, "postal codes were not located since the request is denied due to lack of an API key")

<[REQUEST_DENIED] Google - Geocode [empty]>
<[REQUEST_DENIED] Google - Geocode [empty]>
<[REQUEST_DENIED] Google - Geocode [empty]>
<[REQUEST_DENIED] Google - Geocode [empty]>
<[REQUEST_DENIED] Google - Geocode [empty]>
<[REQUEST_DENIED] Google - Geocode [empty]>
<[REQUEST_DENIED] Google - Geocode [empty]>
<[REQUEST_DENIED] Google - Geocode [empty]>
<[REQUEST_DENIED] Google - Geocode [empty]>
<[REQUEST_DENIED] Google - Geocode [empty]>
<[REQUEST_DENIED] Google - Geocode [empty]>
<[REQUEST_DENIED] Google - Geocode [empty]>
<[REQUEST_DENIED] Google - Geocode [empty]>
<[REQUEST_DENIED] Google - Geocode [empty]>
<[REQUEST_DENIED] Google - Geocode [empty]>
<[REQUEST_DENIED] Google - Geocode [empty]>
<[REQUEST_DENIED] Google - Geocode [empty]>
<[REQUEST_DENIED] Google - Geocode [empty]>
<[REQUEST_DENIED] Google - Geocode [empty]>
<[REQUEST_DENIED] Google - Geocode [empty]>
<[REQUEST_DENIED] Google - Geocode [empty]>
<[REQUEST_DENIED] Google - Geocode [empty]>
<[REQUEST_DENIED] Google - Geoco

Then, we do the exercise with the csv file

In [9]:
loc=pd.read_csv("http://cocl.us/Geospatial_data")
loc=loc.rename(columns={"Postal Code": "Postal code"})
loc

Postal code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

And then we concatenate the two dataframes

In [10]:
df = pd.merge(db, loc, on='Postal code')
df

Postal code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                           Regent Park / Harbourfront  43.654260 -79.360636  
3                    Lawrence Manor / Lawrence Heights  43.718518 -79.464763  
4         Queen's Park / Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98     The Kingsway / Montgomery Road / Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100              Business reply mail Processing CentrE  43.662744 -79.321558  
101  Old Mill South / King's Mill Park / Sunnylea /...  43.636258 -79.498509  
102  Mimico NW / The Queensway West / South of Bloo...  43.628841 -79.520999  

[103 rows x 5 columns]

### Third Section!

We start by updating the foursquare info

In [11]:
CLIENT_ID = 'BZEGAEE1G0YMB0MQKCJ4VA4DBHZ4KCDAXSYZL5IPT1RVKFFB'
CLIENT_SECRET = 'RQSTQNRRCXQNLDN4UBKV213J3UO53Y4OZGSHQRBQEHZYZB5U'
VERSION = '20200417'

print('Your credentails haave been saved:')
print('CLIENT_ID is saved')
print('CLIENT_SECRET is saved')
print('VERSION is', VERSION)

Your credentails haave been saved:
CLIENT_ID is saved
CLIENT_SECRET is saved
VERSION is 20200417


We then use the function to get the vanues in each neighborhood. 

In [12]:
import requests
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        venues_list.append([(
             name, 
             lat, 
             lng, 
             v['venue']['name'], 
             v['venue']['location']['lat'], 
             v['venue']['location']['lng'],  
             v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
        'Neighborhood Latitude', 
        'Neighborhood Longitude', 
        'Venue', 
        'Venue Latitude', 
        'Venue Longitude', 
        'Venue Category']
    
    return(nearby_venues)

And then, to define the radius of search, we calculate the mean neighborhood area by dividing the whole Toronto area (243.3 mi²) into the 103 neighborhoods we will search (2.36 mi²) and then, we obtain the radius (0.86 mi) and then we set the limit of search to 100 venues. The radius defines to avoid double counting venues in different neighborhoods.

In [13]:
radius= 0.866
LIMIT=100

Then, we do the search for each neighborhood

In [14]:
toronto_venues = getNearbyVenues(names=df['Neighborhood'],
                                   latitudes=df['Latitude'],
                                   longitudes=df['Longitude']
                                  )

Parkwoods
Victoria Village
Regent Park / Harbourfront
Lawrence Manor / Lawrence Heights
Queen's Park / Ontario Provincial Government
Islington Avenue
Malvern / Rouge
Don Mills
Parkview Hill / Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park / Princess Gardens / Martin Grove / Islington / Cloverdale
Rouge Hill / Port Union / Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate / Bloordale Gardens / Old Burnhamthorpe / Markland Wood
Guildwood / Morningside / West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor / Wilson Heights / Downsview North
Thorncliffe Park
Richmond / Adelaide / King
Dufferin / Dovercourt Village
Scarborough Village
Fairview / Henry Farm / Oriole
Northwood Park / York University
East Toronto
Harbourfront East / Union Station / Toronto Islands
Little Portugal / Trinity
Kennedy Park / Ionview / East Birchmount Park
Bayview Village
Do

We then analyze how many venues did we find.

In [15]:
print(toronto_venues.shape)
toronto_venues.head()

(2124, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2         Parkwoods              43.753259              -79.329656   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude     Venue Category  
0         Brookbanks Park       43.751976       -79.332140               Park  
1         GTA Restoration       43.753396       -79.333477    Fireworks Store  
2           Variety Store       43.751974       -79.333114  Food & Drink Shop  
3  Victoria Village Arena       43.723481       -79.315635       Hockey Arena  
4             Tim Hortons       43.725517       -79.313103        Coffee Shop

Similarly, we check how many venues we find on each neighborhood andwe organize them according to the number of venues. This would be helpful if we are interested in knowing what neighborhoods are more comercial or residential

In [16]:
number=pd.DataFrame(toronto_venues.groupby('Neighborhood').count()['Venue'])
venues = pd.merge(df, number, on='Neighborhood', how='left')
venues=venues.fillna(0)
venues=venues.sort_values(by=['Venue']).reset_index(drop=True)
venues

Postal code           Borough  \
0           M1X       Scarborough   
1           M2M        North York   
2           M9A         Etobicoke   
3           M9B         Etobicoke   
4           M1J       Scarborough   
..          ...               ...   
98          M5J  Downtown Toronto   
99          M5B  Downtown Toronto   
100         M5K  Downtown Toronto   
101         M5X  Downtown Toronto   
102         M5L  Downtown Toronto   

                                          Neighborhood   Latitude  Longitude  \
0                                          Upper Rouge  43.836125 -79.205636   
1                             Willowdale / Newtonbrook  43.789053 -79.408493   
2                                     Islington Avenue  43.667856 -79.532242   
3    West Deane Park / Princess Gardens / Martin Gr...  43.650943 -79.554724   
4                                  Scarborough Village  43.744734 -79.239476   
..                                                 ...        ...        ...   
98   Harbourfront East / Union Station / Toronto Is...  43.640816 -79.381752   
99                            Garden District, Ryerson  43.657162 -79.378937   
100          Toronto Dominion Centre / Design Exchange  43.647177 -79.381576   
101            First Canadian Place / Underground city  43.648429 -79.382280   
102                    Commerce Court / Victoria Hotel  43.648198 -79.379817   

     Venue  
0      0.0  
1      0.0  
2      0.0  
3      0.0  
4      1.0  
..     ...  
98   100.0  
99   100.0  
100  100.0  
101  100.0  
102  100.0  

[103 rows x 6 columns]

In [17]:
n=venues["Venue"].unique().shape[0]
vector= pd.DataFrame(range(1, n + 1 ,1))
venues['marker_color'] = pd.cut(venues['Venue'], bins=n, 
                              labels=vector[0])


import matplotlib.cm as cm
import matplotlib.colors as colors
import folium
import numpy as np

latitude=43.6532
longitude=-79.3832

map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
colors_array = cm.rainbow(np.linspace(0, 1, n))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(venues['Latitude'], venues['Longitude'], venues['Neighborhood'], venues['marker_color']):
    label = folium.Popup(str(poi) + ', Venues:' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

We found neighborhoods that did not have any venues, and some with several. The purple colors are more residential areas and the more red areas are the ones who are more popular and commercial.

Now we continue to analize the type of commerce in the areas. And transfor the data to find the proportion of venues in each neighborhood.

In [18]:
ven_dum = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
ven_dum['Name']=toronto_venues['Neighborhood'] 
ven_dum

fixed_columns = [ven_dum.columns[-1]] + list(ven_dum.columns[:-1])
ven_dum = ven_dum[fixed_columns]
ven_dum.head()

Name  Accessories Store  Airport  Airport Food Court  \
0         Parkwoods                  0        0                   0   
1         Parkwoods                  0        0                   0   
2         Parkwoods                  0        0                   0   
3  Victoria Village                  0        0                   0   
4  Victoria Village                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  ...  Train Station  \
0                    0             0  ...              0   
1                    0             0  ...              0   
2                    0             0  ...              0   
3                    0             0  ...              0   
4                    0             0  ...              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Video Store  \
0                              0                 0            0   
1                              0                 0            0   
2                              0                 0            0   
3                              0                 0            0   
4                              0                 0            0   

   Vietnamese Restaurant  Warehouse Store  Wine Bar  Wings Joint  \
0                      0                0         0            0   
1                      0                0         0            0   
2                      0                0         0            0   
3                      0                0         0            0   
4                      0                0         0            0   

   Women's Store  Yoga Studio  
0              0            0  
1              0            0  
2              0            0  
3              0            0  
4              0            0  

[5 rows x 269 columns]

Then we create the dummy variables

In [19]:
dum_grouped = ven_dum.groupby('Name').mean().reset_index()
dum_grouped

Name  Accessories Store  \
0                                           Agincourt                0.0   
1                             Alderwood / Long Branch                0.0   
2   Bathurst Manor / Wilson Heights / Downsview North                0.0   
3                                     Bayview Village                0.0   
4                  Bedford Park / Lawrence Manor East                0.0   
..                                                ...                ...   
89                                         Willowdale                0.0   
90                                             Woburn                0.0   
91                                   Woodbine Heights                0.0   
92                          York Mills / Silver Hills                0.0   
93                                    York Mills West                0.0   

    Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0       0.0                 0.0           0.0             0.0   
1       0.0                 0.0           0.0             0.0   
2       0.0                 0.0           0.0             0.0   
3       0.0                 0.0           0.0             0.0   
4       0.0                 0.0           0.0             0.0   
..      ...                 ...           ...             ...   
89      0.0                 0.0           0.0             0.0   
90      0.0                 0.0           0.0             0.0   
91      0.0                 0.0           0.0             0.0   
92      0.0                 0.0           0.0             0.0   
93      0.0                 0.0           0.0             0.0   

    Airport Service  Airport Terminal  American Restaurant  Antique Shop  ...  \
0               0.0               0.0             0.000000           0.0  ...   
1               0.0               0.0             0.000000           0.0  ...   
2               0.0               0.0             0.000000           0.0  ...   
3               0.0               0.0             0.000000           0.0  ...   
4               0.0               0.0             0.038462           0.0  ...   
..              ...               ...                  ...           ...  ...   
89              0.0               0.0             0.000000           0.0  ...   
90              0.0               0.0             0.000000           0.0  ...   
91              0.0               0.0             0.000000           0.0  ...   
92              0.0               0.0             0.000000           0.0  ...   
93              0.0               0.0             0.000000           0.0  ...   

    Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0             0.0                            0.0               0.0   
1             0.0                            0.0               0.0   
2             0.0                            0.0               0.0   
3             0.0                            0.0               0.0   
4             0.0                            0.0               0.0   
..            ...                            ...               ...   
89            0.0                            0.0               0.0   
90            0.0                            0.0               0.0   
91            0.0                            0.0               0.0   
92            0.0                            0.0               0.0   
93            0.0                            0.0               0.0   

    Video Store  Vietnamese Restaurant  Warehouse Store  Wine Bar  \
0      0.000000                  0.000              0.0       0.0   
1      0.000000                  0.000              0.0       0.0   
2      0.052632                  0.000              0.0       0.0   
3      0.000000                  0.000              0.0       0.0   
4      0.000000                  0.000              0.0       0.0   
..          ...                    ...              ...       ...   
89     0.000000                  0.025              0.0   

And find the most common venue for each neighborhood.

In [20]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = dum_grouped['Name']

for ind in np.arange(dum_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(dum_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                          Agincourt                Lounge   
1                            Alderwood / Long Branch           Pizza Place   
2  Bathurst Manor / Wilson Heights / Downsview North           Coffee Shop   
3                                    Bayview Village                  Café   
4                 Bedford Park / Lawrence Manor East        Sandwich Place   

  2nd Most Common Venue      3rd Most Common Venue 4th Most Common Venue  \
0        Breakfast Spot  Latin American Restaurant          Skating Rink   
1                   Gym                   Pharmacy        Sandwich Place   
2                  Bank        Fried Chicken Joint           Bridal Shop   
3                  Bank         Chinese Restaurant   Japanese Restaurant   
4    Italian Restaurant                 Restaurant           Coffee Shop   

         5th Most Common Venue 6th Most Common Venue    7th Most Common Venue  \
0  Eastern European Restaurant               Dog Run         Doner Restaurant   
1                          Pub    Athletics & Sports                     Pool   
2               Sandwich Place            Restaurant                    Diner   
3                  Yoga Studio               Dog Run         Doner Restaurant   
4             Sushi Restaurant           Pizza Place  Comfort Food Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Donut Shop             Drugstore    Dumpling Restaurant  
1          Skating Rink           Coffee Shop      Convenience Store  
2        Ice Cream Shop           Supermarket       Sushi Restaurant  
3            Donut Shop             Drugstore    Dumpling Restaurant  
4       Thai Restaurant             Juice Bar   Fast Food Restaurant

We custer the neighborhoods in 10 groups

In [21]:
from sklearn.cluster import KMeans

kclusters = 10

grouped_clustering = dum_grouped.drop('Name', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3, 1, 1, 1, 1, 1, 1, 6, 1, 1, 1, 7,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 5, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 7, 0, 1, 4, 1, 1, 1, 8, 1, 1, 1, 1, 1, 1, 1, 1, 0,
       1, 1, 1, 2, 4, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 3, 1, 1, 1, 1, 1, 0,
       1, 1, 1, 1, 9, 1])

The results show that most neighborhoods are part of the same cluster, and there are a few that do not have a cluster, since they did not have any venues found. Now, we join the data about the neighborhoods, the clusters and the most common venues. We then add the 11th group (cluster 10) refering to the group with no venues.

In [22]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

klusterdf = df
klusterdf= klusterdf.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')
klusterdf=klusterdf.fillna(10)
klusterdf.head()

Postal code           Borough                                  Neighborhood  \
0         M3A        North York                                     Parkwoods   
1         M4A        North York                              Victoria Village   
2         M5A  Downtown Toronto                    Regent Park / Harbourfront   
3         M6A        North York             Lawrence Manor / Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             1.0                  Park   
1  43.725882 -79.315572             1.0         Grocery Store   
2  43.654260 -79.360636             1.0           Coffee Shop   
3  43.718518 -79.464763             1.0        Clothing Store   
4  43.662301 -79.389494             1.0           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
0       Fireworks Store     Food & Drink Shop            Event Space   
1           Coffee Shop            Nail Salon  Portuguese Restaurant   
2                Bakery                   Pub                   Park   
3         Women's Store     Accessories Store            Coffee Shop   
4      Sushi Restaurant                 Diner          Burrito Place   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0  Ethiopian Restaurant   Empanada Restaurant     Electronics Store   
1          Hockey Arena     Electronics Store   Empanada Restaurant   
2               Theater                  Café        Breakfast Spot   
3           Event Space            Shoe Store    Miscellaneous Shop   
4                  Bank                   Bar              Beer Bar   

         8th Most Common Venue  9th Most Common Venue 10th Most Common Venue  
0  Eastern European Restaurant         Farmers Market    Dumpling Restaurant  
1  Eastern European Restaurant    Dumpling Restaurant     Dim Sum Restaurant  
2                   Restaurant         Chocolate Shop    Distribution Center  
3       Furniture / Home Store  Vietnamese Restaurant               Boutique  
4                          Spa     Italian Restaurant           Burger Joint

Then, we create the map to viasualize the data.

In [23]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)
kclusters=11
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

markers_colors = []
for lat, lon, poi, cluster in zip(klusterdf['Latitude'], klusterdf['Longitude'], klusterdf['Neighborhood'], klusterdf['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster)-1],
        fill=True,
        fill_color=rainbow[int(cluster)-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Finally, the map is showing us that the clustering may not be very good due to the high ammount of categories of the venues. Additionally, this analysis does not discriminate in the ammount and venues, and therefore does not differentiate in between residential and commercial areas.